### PR TITLE
Make sure correct volume mounts are created

### LINKED ISSUE #309: The operator can generate incorrect volume mounts
Note that this issue surface while I was working with a recent build from master, not from a previously released version.

There are a couple problems that can happen you have a `.spec.podTemplateSpec` that includes an initContainer with volume mounts. Here is an example manifest with which I reproduced the issues:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: medusa-test
  serverType: cassandra
  serverVersion: 3.11.7
  managementApiAuth:
    insecure: {}
  size: 3
  allowMultipleNodesPerWorker: true
  config:
    jvm-options:
      initial_heap_size: "1024m"
      max_heap_size: "1024m"
  resources:
    requests:
      cpu: 1
      memory: 2Gi
    limits:
      cpu: 1
      memory: 2Gi
  podTemplateSpec:
    spec:
      initContainers:
        - name: medusa-restore
          image: jsanda/medusa:grpc-dev
          imagePullPolicy: IfNotPresent
          env:
            - name: MEDUSA_MODE
              value: RESTORE
          volumeMounts:
            - name: medusa-gcp-key
              mountPath: /etc/medusa-secrets
              subPath: keys
            - name: medusa-config
              mountPath: /etc/medusa/medusa.ini
              subPath: medusa.ini
            - name: server-config
              mountPath: /etc/cassandra
            - mountPath: /var/lib/cassandra
              name: server-data
      containers:
        - name: cassandra
          env:
          - name: JVM_EXTRA_OPTS
            value: -javaagent:/etc/cassandra/jolokia-jvm-1.6.2-agent.jar=port=7373,host=localhost
      volumes:
        - name: medusa-gcp-key
          secret:
            secretName: medusa-gcp-key
        - name: medusa-config
          configMap:
            name: medusa-config
            items:
            - key: medusa.ini
              path: medusa.ini
  storageConfig:
    cassandraDataVolumeClaimSpec:
      storageClassName: standard
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
```
I declare two additional volumes, `medusa-gcp-key` and `medusa-config`. The `medusa-restore` initContainer declares volume mounts for them as well as for `server-config` and `server-data`. The operator creates the latter two.

Here is the relevant part of the stateful spec that the operator generates:

```yaml
      containers:
      - env:
        - name: JVM_EXTRA_OPTS
          value: -javaagent:/etc/cassandra/jolokia-jvm-1.6.2-agent.jar=port=7373,host=localhost
        - name: DS_LICENSE
          value: accept
        - name: DSE_AUTO_CONF_OFF
          value: all
        - name: USE_MGMT_API
          value: "true"
        - name: MGMT_API_EXPLICIT_START
          value: "true"
        - name: DSE_MGMT_EXPLICIT_START
          value: "true"
        image: datastax/cassandra-mgmtapi-3_11_7:v0.1.13
        imagePullPolicy: IfNotPresent
        lifecycle:
          preStop:
            exec:
              command:
              - wget
              - --output-document
              - /dev/null
              - --no-check-certificate
              - --post-data=''
              - http://localhost:8080/api/v0/ops/node/drain
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /api/v0/probes/liveness
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 15
          periodSeconds: 15
          successThreshold: 1
          timeoutSeconds: 1
        name: cassandra
        # ports:  omitted for brevity       
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /api/v0/probes/readiness
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 20
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 2Gi
          requests:
            cpu: "1"
            memory: 2Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /config
          name: server-config
        - mountPath: /etc/medusa-secrets
          name: medusa-gcp-key
          subPath: keys
        - mountPath: /etc/medusa/medusa.ini
          name: medusa-config
          subPath: medusa.ini
        - mountPath: /var/lib/cassandra
          name: server-data
        - mountPath: /var/log/cassandra
          name: server-logs
        - mountPath: /var/lib/cassandra
          name: server-data
        - mountPath: /etc/encryption/
          name: encryption-cred-storage
```
There are two issues. First, we have a duplicate volumeMount for `server-data`. This results in an invalid spec and the statefulset controller does not create any pods.

The second issue is that the volumeMounts that I specified for my initContainer have been added to the `cassandra` container.

I am able to work around the duplicate mount by removing the `server-data` mount from my initContainer.

I think the problem stems from [here](https://github.com/datastax/cass-operator/blob/master/operator/pkg/reconciliation/construct_podtemplatespec.go#L365). Volume mounts declared for any initContainers are added to the `volumeDefaults` slice. Those get added to the `cassandra` container. This causes the `cassandra` to end up with volumeMounts that it should not have.

The duplicate volumeMount happens [here](https://github.com/datastax/cass-operator/blob/master/operator/pkg/reconciliation/construct_podtemplatespec.go#L376). The duplicate mount could be avoided by first checking to see if `/var/lib/cassandra` is already in the list of mounts. I think a better solution however is to *not* take volumeMounts declared in initContainers and add them to the `cassandra` container.

### PR DESCRIPTION
**Pull request by jsanda**
Fixes #309

### CODE REVIEW

File: operator/pkg/reconciliation/construct_podtemplatespec_test.go [hunk 1]
```diff
@@ -410,6 +410,280 @@ func TestCassandraDatacenter_buildPodTemplateSpec_initcontainers_merge(t *testin
 	}
 }
 
+func TestCassandraDatacenter_buildPodTemplateSpec_add_initContainer_after_config_initContainer(t *testing.T) {
+	// When adding an initContainer with podTemplate spec it will run before
+	// the server config initContainer by default. This test demonstrates and
+	// verifies how to specify the initContainer to run after the server config
+	// initContainer.
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec_test.go:417]: 💯

File: operator/pkg/reconciliation/construct_podtemplatespec_test.go [hunk 2]
```diff
@@ -410,6 +410,280 @@ func TestCassandraDatacenter_buildPodTemplateSpec_initcontainers_merge(t *testin
...
+
+type VolumeMountMatcher func(volumeMount corev1.VolumeMount) bool
+
+type VolumeMatcher func(volume corev1.Volume) bool
+
+func volumeNameMatcher(name string) VolumeMatcher {
+	return func(volume corev1.Volume) bool {
+		return volume.Name == name
+	}
+}
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec_test.go:652]: I'm good with this, but I feel obliged to point out Rob Pike's repo here -> https://github.com/robpike/filter

File: operator/pkg/reconciliation/construct_podtemplatespec_test.go [hunk 3]
```diff
@@ -436,7 +710,7 @@ func TestCassandraDatacenter_buildPodTemplateSpec_labels_merge(t *testing.T) {
 }
 
 // A volume added to ServerConfigContainerName should be added to all built-in containers
-func TestCassandraDatacenter_buildPodTemplateSpec_propagate_volumes(t *testing.T) {
+func TestCassandraDatacenter_buildPodTemplateSpec_do_not_propagate_volumes(t *testing.T) {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec_test.go:713]: 🤣

**Review by jimdickinson — APPROVED** [review]
lgtm - thank you for the great tests and comments!

File: operator/pkg/reconciliation/construct_podtemplatespec_test.go (hunk 2 above)
Comment (jsanda) [reply operator/pkg/reconciliation/construct_podtemplatespec_test.go:652]: Do you want me to update the code to use that? Happy to do so. I promise you I do not have NIH syndrome :)

Comment (jimdickinson) [reply operator/pkg/reconciliation/construct_podtemplatespec_test.go:652]: to get rid of all the one line functions and functions abstracting a for loop? no it reads cleanly here, and this is test code, if it's working right now let's keep it as-is

**Review by jimdickinson — APPROVED** [review]
lgtm

**Pull request merged**